controllers/main.go

```diff
@@ -137,55 +139,67 @@ func NewMainController(params *chaincfg.Params, adminIPs []string, baseURL strin
 	return mc, nil
 }
 
-// API is the main frontend that handles all API requests.
-// XXX This is very hacky.  It can be made more elegant by improving or
-// re-writing the middleware and switching to a more modern framework
-// such as goji2 or something else that utilizes the context package/built-in.
-// This would make it much easier to share code between the web UI and JSON
-// API with less duplication.
-func (controller *MainController) API(c web.C, r *http.Request) (string, int) {
-	// we expect /api/vX.YZ/command
-	URIparts := strings.Split(r.RequestURI, "/")
-	if len(URIparts) != 4 {
-		return "{\"status\": \"error\"," +
-			"\"message\": \"invalid API request\"}\n", http.StatusOK
+func (c *MainController) APIInvalidHandler(w http.ResponseWriter, r *http.Request) {
+	resp := poolapi.Response{Status: "error",
+		Message: "invalid API version",
 	}
-	version := URIparts[2]
-	command := URIparts[3]
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(http.StatusNotFound)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Warnf("JSON encode error: %v", err)
+	}
+}
 
-	if version != "v0.1" {
-		return "{\"status\": \"error\"," +
-			"\"message\": \"invalid API version\"}\n", http.StatusOK
+func APIResponseHandler(resp *poolapi.Response, code int, w http.ResponseWriter,
+	r *http.Request) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(code)
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Warnf("JSON encode error: %v", err)
 	}
+}
+
+// APIHandler is the main frontend that handles all API requests.
+func (controller *MainController) APIHandler(c web.C, w http.ResponseWriter,
+	r *http.Request) {
+	command := c.URLParams["command"]
+
+	httpStatus := http.StatusOK
+	var aresp *poolapi.Response
 
 	switch r.Method {
 	case "GET":
 		switch command {
-		case "getPurchaseInfo":
+		case "getpurchaseinfo":
```

Comment:
and fwiw, i'm ok with making the api more consistent by doing this, but please do it in a different commit with a bump to the version.  refactoring prs shouldn't change observable behavior.